include/tscpp/util/Hash.h

```diff
@@ -0,0 +1,153 @@
+/** @file Basic hash function support.
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <cctype>
+#include "tscpp/util/MemSpan.h"
+#include "tscpp/util/TextView.h"
+
+namespace ts
+{
+/** Base protocol class for hash functors.
+ *
+ * Each specific hash function embedded in a hash functor is a subclass of this class and
+ * follows this API. Subclasses should override the return type to return the subclass type.
+ *
+ * The main purpose of this is to allow run time changes in hashing, which is required in various
+ * circumstances.
+ */
+struct HashFunctor {
+  using self_type = HashFunctor; ///< Self reference type.
+
+  /// Pass @a data to the hashing function.
+  virtual self_type &update(std::string_view const &data) = 0;
+
+  /// Finalize the hash function output.
+  virtual self_type & final() = 0;
+
+  /// Reset the hash function state.
+  virtual self_type &clear() = 0;
+
+  /// Get the size of the resulting hash value.
+  virtual size_t size() const = 0;
+
+  /// Copy the result to @a dst.
+  /// @a dst must be at least @c result_size bytes long.
+  /// @return @c true if the result was copied to @a data, @c false otherwise.
+  virtual bool get(MemSpan dst) const = 0;
+
+  virtual ~HashFunctor() = default; ///< Force virtual destructor.
+};
+
+/// A hash function that returns a 32 bit result.
+struct Hash32Functor : HashFunctor {
```

Comment:
OK I looked it up structures, unlike classes, have public inheritance by default.  Or, you could say public and people would not have to look it up.

Reply:
That's an interview question - you shouldn't have to look it up. It's less obscure than "final" being a keyword.

Reply:
Haha too late already hired me.

Reply:
I like struct for that exact reason, it instantly tells me that it's all public, and I don't think we should discourage the use of struct over class.

Reply:
@zwoop I was referring to using default access control for inheritance.  I prefer it explicit, unlike SolidWallOfObscurity.

Reply:
I made another comment above, questioning the use of struct instead of class, when the class has some members with private or protected access control.

Reply:
In this particular case, it's inertia. When I write new code, I tend to follow Leif's approach and only use `struct` for POD things.